### PR TITLE
More Capability APIs

### PR DESCRIPTION
**Pull request by zyga**
This is the last of the branches that lead to the REST API for capabilities.
This pull request depends on #143 so please land that first.

Please check the comments on particular commits for details.

### CODE REVIEW

**Comment by zyga** [comment]
This branch contains all the fixes mentioned in #153 without actually including the REST APIs.

**Comment by chipaca** [comment]
:+1:

File: caps/repo.go [hunk 1]
```diff
@@ -81,6 +81,17 @@ func (r *Repository) HasType(t *Type) bool {
 	return false
 }
 
+// FindTypeByName finds and returns the Type with the given name or nil if
+// it's not found
+func (r *Repository) FindTypeByName(name string) *Type {
```
Comment (niemeyer) [inline caps/repo.go]: s/FindTypeByName/Type/? :-)

File: caps/types.go [hunk 2]
```diff
@@ -65,3 +63,18 @@ func (t *Type) Validate(c *Capability) error {
 	}
 	return nil
 }
+
+// TypeLookupFn aids in looking up a Type by name
+type TypeLookupFn func(name string) *Type
```
Comment (niemeyer) [inline caps/types.go]: Is that a left over?  Doesn't seem worth it or relevant to the branch.

File: caps/types.go [hunk 3]
```diff
@@ -65,3 +63,18 @@ func (t *Type) Validate(c *Capability) error {
...
+// MarshalJSON encodes a Type object as the name of the type.
+func (t *Type) MarshalJSON() ([]byte, error) {
+	return json.Marshal(t.Name)
+}
+
+// UnmarshalJSON decodes the name of a Type object.
+// NOTE: In the future, when more properties are added, those properties will
+// not be decoded and will be left over as empty values.
+func (t *Type) UnmarshalJSON(data []byte) error {
+	return json.Unmarshal(data, &t.Name)
```
Comment (niemeyer) [inline caps/types.go]: Hmm.. seems okay for now, but it's a bit eye-catching. It's loading a broken type. I think we'll need to return to this later.

File: caps/types.go [hunk 4]
```diff
@@ -65,3 +63,18 @@ func (t *Type) Validate(c *Capability) error {
 	}
 	return nil
 }
+
+// TypeLookupFn aids in looking up a Type by name
+type TypeLookupFn func(name string) *Type
+
+// MarshalJSON encodes a Type object as the name of the type.
+func (t *Type) MarshalJSON() ([]byte, error) {
+	return json.Marshal(t.Name)
```
Comment (niemeyer) [inline caps/types.go]: Not clear yet if that's what we really want here, or whether we want an actual objects with some fields. Seems fine for now.

File: caps/types.go (hunk 3 above)
Comment (zyga) [inline caps/types.go]: This is the best way we've found to go ahead without CapabilityRepr. I'm open to ideas on how to handle it better without context-sensitive unmarshaller.

File: caps/types.go (hunk 4 above)
Comment (zyga) [inline caps/types.go]: I think we want just the name as the type will have extra initialized, validated structures that we don't want to handle twice on the daemon side. Alternatively, we can hold those separate and just keep the type Type string; alias

**Comment by zyga** [comment]
Corrected

**Comment by niemeyer** [comment]
LGTM. As mentioned online, I think we'll eventually need an error result on the Type method.. while it's just a matter of found/not found, seems fine not to have one, but soon we might be loading those from external sources, which means we'll want to communicate error cases better. Fine to wait, though.

**Pull request merged**